app/src/main/java/app/revanced/integrations/shared/GmsCoreSupport.java

```diff
@@ -81,40 +89,73 @@ public static void checkGmsCore(Context context) {
                 open(getGmsCoreDownload());
                 return;
             }
 
-            // Check if GmsCore is whitelisted from battery optimizations.
-            var powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
-            if (!powerManager.isIgnoringBatteryOptimizations(GMS_CORE_PACKAGE_NAME)) {
-                Logger.printDebug(() -> "GmsCore is not whitelisted from battery optimizations");
-                showToastOrDialog(context,
-                        "gms_core_toast_not_whitelisted_message",
-                        "gms_core_dialog_not_whitelisted_using_battery_optimizations_message",
-                        DONT_KILL_MY_APP_LINK);
-                return;
-            }
-
             // Check if GmsCore is running in the background.
             try (var client = context.getContentResolver().acquireContentProviderClient(GMS_CORE_PROVIDER)) {
                 if (client == null) {
-                    Logger.printDebug(() -> "GmsCore is not running in the background");
-                    showToastOrDialog(context,
-                            "gms_core_toast_not_whitelisted_message",
+                    Logger.printInfo(() -> "GmsCore is not running in the background");
+
+                    showBatteryOptimizationToastOrDialog(context,
                             "gms_core_dialog_not_whitelisted_not_allowed_in_background_message",
-                            DONT_KILL_MY_APP_LINK);
+                            "gms_core_dialog_open_website_text",
+                            (dialog, id) -> open(DONT_KILL_MY_APP_LINK),
+                            null);
+                    return;
                 }
             }
+
+            // Check if GmsCore is whitelisted.
+            if (batteryOptimizationsTurnedOn(context)) {
+                Logger.printInfo(() -> "GmsCore is not whitelisted from battery optimizations");
+                showBatteryOptimizationToastOrDialog(context,
+                        "gms_core_dialog_not_whitelisted_using_battery_optimizations_message",
+                        "gms_core_dialog_continue_text",
+                        (dialog, id) -> openGmsDisableBatteryOptimizationsIntent((Activity) context),
+                        dialog -> checkIfUserAllowedTurningOffBatteryOptimizations(context));
+            }
         } catch (Exception ex) {
             Logger.printException(() -> "checkGmsCore failure", ex);
         }
     }
 
+    private static void checkIfUserAllowedTurningOffBatteryOptimizations(Context context) {
+        // Check if the user actually turned off power optimizations for GmsCore.
+        // This can be done by patching the Activity and overriding the
+        // onActivityResult() method, but that only works for YT
+        // and not YT Music because Music is not hooked to an Activity.
+        // Instead, check back in 20 seconds and show a toast if it's still not enabled.
```

Comment:
Probably this should be worded generalized. If at all, YouTube and YouTube Music should be mentioned as example cases, since the patch is generalized for any app.